cdk/kittyhawk/src/ingress.ts

```diff
@@ -13,55 +13,112 @@ export interface IngressProps {
   /**
      * A list of host rules used to configure the Ingress.
      *
-     * @default undefined
+     * @default []
+     */
+  readonly rules: NonEmptyArray<HostRules>;
+
+  /**
+     * A key/value map of annotations to customize the Ingress (do path prefix routing, etc.).
+     *
+     * @default {}
      */
-  readonly ingress?: { host: string; paths: string[] }[];
+  readonly annotations?: { [key: string]: string };
+}
+
+export interface HostRules {
+  /**
+   * The domain for the ingress.
+   */
+  readonly host: string;
+
+  /**
+   * Paths on the domain that the application should be available.
+   */
+  readonly paths: string[];
+
+  /**
+   * If the host is a subdomain.
+   */
+  readonly isSubdomain?: boolean;
 }
 
 export class Ingress extends Construct {
   constructor(scope: Construct, appname: string, props: IngressProps) {
     super(scope, `ingress-${appname}`);
 
-    const port = props.port || 80;
-    const ingress = props.ingress;
 
-    if (ingress) {
-      let tls = ingress.map(h => {
-        // Regex to compute the apex domain
-        const apex_domain = h.host.match(/[\w-]+\.[\w]+$/g);
-        if (apex_domain) {
-          const host_string = apex_domain[0].split('.').join('-').concat('-tls');
-          return { hosts: [h.host], secretName: host_string };
-        } else {throw new Error(`Ingress construction failed: apex domain regex failed on ${h}`);}
-      });
+    const fullConfig: Required<IngressProps> = {
+      port: 80,
+      annotations: props.annotations ?? {},
+      ...props,
+    };
+
+    const tls = props.rules.map(h => {
+      const hostString: string = `${domainToCertName(h.host, h.isSubdomain ?? false)}-tls`;
+      return { hosts: [h.host], secretName: hostString };
+    });
 
-      let rules = ingress.map(h => {
-        return {
-          host: h.host,
-          http: {
-            paths: h.paths.map(path => {
-              return {
-                path: path,
-                backend: {
-                  serviceName: appname,
-                  servicePort: IntOrString.fromNumber(port),
-                },
-              };
-            }),
+    const rules: IngressRule[] = props.rules.map(h => ({
+      host: h.host,
+      metadata: {
+        annotations: props.annotations,
+      },
+      http: {
+        paths: h.paths.map(path => ({
+          path: path,
+          pathType: 'Prefix',
+          backend: {
+            service: {
+              name: appname,
+              port: {
+                number: fullConfig.port,
+              },
+            },
           },
-        };
-      });
+        })),
+      },
+    }),
+    );
+
+    new IngressApiObject(this, `ingress-${appname}`, {
```

Comment:
@brandonwang1 Why is the construct of this API object `this`, as opposed to `scope`?

Reply:
Actually just kinda confused when this and scope is used in general

Reply:
its been a long time since ive looked at this so details may not be 100% right _but_: scope is used so cdk8s can keep track of the [construct tree](https://cdk8s.io/docs/latest/getting-started/#apps-charts) that is being generated (its used to keep track of everything during synthesis), so if you look in chart.ts the scope is first created when `const app = new App();` is called and it's passed down to all the constructs through scope. `this` references the object (so in ingress.ts it references the Ingress class). Then since Ingress class gets its scope from the scope parameter passed through the constructor so if we use `this` on line 83 i think (not totally sure here) it just makes IngressApiObject a child of the Ingress construct in the construct tree. I imagine using scope here would work too but its slightly better to make the relationship clear between IngressApiObject and Ingress rather than just making IngressApiObject a child of `scope` (really `app`) (can you test if it works both ways just to make sure?)